cds_books/migrator/records.py

```diff
@@ -0,0 +1,196 @@
+            UnexpectedValue: migration_exception_handler,
+            MissingRequiredField: migration_exception_handler,
+            ManualMigrationRequired: migration_exception_handler,
+        }
+
+        if self.source_type == 'marcxml':
+            marc_record = create_record(data['marcxml'])
+            try:
+                val = self.dojson_model.do(
+                    marc_record, exception_handlers=exception_handlers)
+                missing = self.dojson_model.missing(marc_record)
+                if missing:
+                    raise LossyConversion(missing=missing)
+                update_access(val, self.collection_access)
+                return dt, val
+            except LossyConversion as e:
+                raise e
+            except Exception as e:
+                current_app.logger.error(
+                    'Impossible to convert to JSON {0} - {1}'.format(
+                        e, marc_record))
+                raise e
+        else:
+            val = data['json']
+
+            # Calculate the _access key
+            update_access(val, self.collection_access)
+            return dt, val
+
+    def prepare_revisions(self):
+        """Prepare data.
+
+        We don't convert intermediate versions to JSON to avoid conversion
+        errors and get a lossless version migration.
+
+        If the revisions is the last one, an error will be generated if the
+        final translation is not complete.
+        """
+        # Prepare revisions
+        self.revisions = []
+
+        it = [self.data['record'][0]] if self.latest_only \
+            else self.data['record']
+
+        for i in it[:-1]:
+            self.revisions.append(self._prepare_intermediate_revision(i))
+
+        self.revisions.append(self._prepare_final_revision(it[-1]))
+
+
+class CDSParentRecordDump(RecordDump):
+
+    def _prepare_revision(self, data):
+        """Prepare data."""
+        dt = datetime.datetime.now()
+
+        val = data
+        val['_collections'] = self.data.get('collections', [])
+
+        return dt, val
+
+    def prepare_revisions(self):
+        self.revisions = [self.data]
+
+
+class CDSParentRecordDumpLoader(RecordDumpLoader):
+
+    @classmethod
+    def create(cls, dump, model=Record):
+        """Create record based on dump."""
+        record = cls.create_record(dump, model)
+        return record
+
+    @classmethod
+    @disable_timestamp
+    def create_record(cls, dump, model=Record):
+        """Create a new record from dump."""
+        # Reserve record identifier, create record and recid pid in one
+        # operation.
+        record_uuid = uuid.uuid4()
+        provider = model._pid_provider.create(
+            object_type='rec',
+            object_uuid=record_uuid,
+        )
+        dump[model.pid_field] = provider.pid.pid_value
+        record = model.create(dump, record_uuid)
+        record.model.created = datetime.datetime.now()
+        record.commit()
+        return record
```

Comment:
Is there any indexing happening at some point?

Reply:
Indexing has now been implemented for migrating parents, documents and relations.